MODELIX-914 Remove in-memory query cache and support querying multiple large branches in parallel

slisson opened the pull request:
TODO description of the PR

## To be verified by reviewers

* [ ] Relevant public API members have been documented
* [ ] Documentation related to this PR is complete
  * [ ] Boundary conditions are documented
  * [ ] Exceptions are documented
  * [ ] Nullability is documented if used
* [ ] Touched existing code has been extended with documentation if missing
* [ ] Code is readable
* [ ] New features and fixed bugs are covered by tests

mhuster23 commented on streams/build.gradle.kts:
```diff
@@ -0,0 +1,62 @@
+
+plugins {
+    `maven-publish`
+    kotlin("multiplatform")
```
We should use `` `modelix-kotlin-multiplatform` `` and remove lines 9-27 to use the multiplatform configuration from the convention.

mhuster23 commented on model-datastructure/src/commonMain/kotlin/org/modelix/model/lazy/BulkQuery.kt:
```diff
@@ -15,200 +15,124 @@
     private val queue: MutableMap<String, QueueElement<out IKVValue>> = LinkedHashMap()
-    private var processing = false
+    private var processing = AtomicBoolean(false)
+    private val processingLock = Any()
```
Can `processingLock` be removed?

mhuster23 commented on light-model-client/src/jvmTest/kotlin/org/modelix/client/light/FakeLightModelServer.kt:
```diff
@@ -79,142 +78,138 @@ class FakeLightModelServer(val client: LocalModelClient, val repositoriesManager
+                                        sendDelta(mergedVersion, replacedIds, message.changeSetId)
+                                        previouslyReplacedIds.putAll(replacedIds)
+                                        // idsKnownByClient.forEach { previouslyReplacedIds.remove(it) }
+                                        // TODO remove all other entries in `previouslyReplacedIds` that were part of the same message
```
Is this TODO still relevant?

slisson replied:
The diff doesn't show it very well, but I only removed the wrapping `repositoriesManager.runWithRepository(repositoryId) { ... }`. I will not change anything else, at least not in this PR.

mhuster23 commented on model-datastructure/src/commonMain/kotlin/org/modelix/model/async/AsyncStoreAsLegacyDeserializingStore.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    override fun prefetch(hash: String) {
+        TODO("Not yet implemented")
```
Was this forgotten? If not, we should have a clearer message.

mhuster23 reviewed (commented):
I read through everything and found a few things. It's hard to give more meaningful feedback since the PR is so big.

languitar commented on docs/global/modules/core/pages/howto/modelql.adoc:
```diff
@@ -195,3 +195,31 @@ val html = query {
+[source,kotlin]
+--
+val nodeId: String
+root.find({ it.descendants() }, { it.nodeReferenceAsString() }, nodeId.asMono())
```
Maybe some inline comments here could help?

```suggestion
root.find(
    // Provides all nodes to index
    { it.descendants() },
    // selects the index key for each node
    { it.nodeReferenceAsString() },
    // The key to search for in the current request
    nodeId.asMono()
)
```

languitar commented on kotlin-utils/src/jvmMain/kotlin/org/modelix/kotlin/utils/ThreadLocal.jvm.kt:
Any reason to include the `.jvm` in the name? Shouldn't this be clear from the source set?

slisson replied:
IntelliJ does that by default.

languitar commented on kotlin-utils/src/jvmMain/kotlin/org/modelix/kotlin/utils/ThreadLocal.jvm.kt:
```diff
@@ -0,0 +1,36 @@
+
+package org.modelix.kotlin.utils
+
+import java.lang.ThreadLocal
+
+actual class ThreadLocal<E> actual constructor(initialValueSupplier: () -> E) {
+
+    private val threadLocal = ThreadLocal.withInitial(initialValueSupplier)
```
```suggestion
import java.lang.ThreadLocal as JvmThreadLocal

actual class ThreadLocal<E> actual constructor(initialValueSupplier: () -> E) {

    private val threadLocal = JvmThreadLocal.withInitial(initialValueSupplier)
```

Could make this easier to read with all the duplicated names.

languitar commented on model-api-gen-runtime/src/commonMain/kotlin/org/modelix/metamodel/UnknownConcept.kt:
```diff
@@ -5,63 +5,13 @@ import org.modelix.model.api.IConcept
-    override fun serialize(): String = "null"
-}
+@Deprecated("use org.modelix.model.api.meta.NullConcept", ReplaceWith("org.modelix.model.api.meta.NullConcept"))
+val NullConcept = org.modelix.model.api.meta.NullConcept
```
Do we still have usages for this one?

slisson replied:
Removed

languitar commented on model-api/src/commonMain/kotlin/org/modelix/model/api/IChildLink.kt:
```diff
@@ -15,9 +15,12 @@
 /**
  * Representation of a parent-child relationship between [IConcept]s.
  */
+@Deprecated("Use IChildLinkReference or IChildLinkDefinition")
```
I can't find an IChildLinkDefinition?

slisson replied:
Added

languitar commented on model-api/src/commonMain/kotlin/org/modelix/model/api/IProperty.kt:
```diff
@@ -15,16 +15,99 @@
 /**
  * Representation of a property within an [IConcept].
  */
+@Deprecated("Use IPropertyReference or IPropertyDefinition")
```
I can't find IPropertyDefinition

slisson replied:
Added

languitar commented on model-api/src/commonMain/kotlin/org/modelix/model/api/IReferenceLink.kt:
```diff
@@ -15,13 +15,91 @@
 /**
  * Representation of a non-containment reference link between [IConcept]s.
  */
+@Deprecated("Use IReferenceLinkReference or IReferenceLinkDefinition")
```
I can't find IReferenceLinkDefinition

slisson replied:
Added

languitar commented on model-api/src/commonMain/kotlin/org/modelix/model/api/IRole.kt:
```diff
@@ -1,11 +1,13 @@
  * An [IRole] is a structural feature of a concept.
  * It can either be an [IProperty] or an [ILink].
  */
+@Deprecated("Use IRoleReference or IRoleDefinition")
```
There's no IRoleDefinition

slisson replied:
Added

languitar commented on model-api/src/commonMain/kotlin/org/modelix/model/api/IRole.kt:
```diff
@@ -37,6 +39,48 @@ interface IRole {
+     */
+    fun getIdOrName(): String
+
+    @Deprecated("use IRoleReference or IRoleDefinition instead of IRole")
```
As above

slisson replied:
Added

languitar reviewed (changes requested):
In the current state, at least one commit needs to be marked as a breaking commit. When using `ModelFacade.toLocalBranch(ModelFacade.newLocalTree())` and then resolving references via ModelQL on the resulting branch, resolution fails due to the lack of an INodeResolutionScope being set in code that worked before.

languitar commented:
One more thing: I've built a docker image of the model-server from this branch. With postgres persistence, the image does not start:

```
[model-server]             | org.apache.ignite.IgniteException: Work directory does not exist and cannot be created: /usr/modelix-model/ignite/work
[model-server]             | 	at org.apache.ignite.internal.util.IgniteUtils.convertException(IgniteUtils.java:1171)
[model-server]             | 	at org.apache.ignite.Ignition.start(Ignition.java:353)
[model-server]             | 	at org.modelix.model.server.store.IgniteStoreClient.<init>(IgniteStoreClient.kt:82)
[model-server]             | 	at org.modelix.model.server.store.IgniteStoreClient.<init>(IgniteStoreClient.kt:41)
[model-server]             | 	at org.modelix.model.server.Main.main(Main.kt:155)
[model-server]             | Caused by: org.apache.ignite.IgniteCheckedException: Work directory does not exist and cannot be created: /usr/modelix-model/ignite/work
[model-server]             | 	at org.apache.ignite.internal.util.IgniteUtils.workDirectory(IgniteUtils.java:10027)
[model-server]             | 	at org.apache.ignite.internal.IgnitionEx$IgniteNamedInstance.initializeConfiguration(IgnitionEx.java:1817)
[model-server]             | 	at org.apache.ignite.internal.IgnitionEx$IgniteNamedInstance.start(IgnitionEx.java:1641)
[model-server]             | 	at org.apache.ignite.internal.IgnitionEx.start0(IgnitionEx.java:1089)
[model-server]             | 	at org.apache.ignite.internal.IgnitionEx.startConfigurations(IgnitionEx.java:983)
[model-server]             | 	at org.apache.ignite.internal.IgnitionEx.start(IgnitionEx.java:889)
[model-server]             | 	at org.apache.ignite.internal.IgnitionEx.start(IgnitionEx.java:828)
[model-server]             | 	at org.apache.ignite.Ignition.start(Ignition.java:350)
[model-server]             | 	... 3 common frames omitted
```

languitar commented:
Trying to resolve non-existing nodes via ModelQl takes ages. It was never fast before, but it feels even longer now. Here's a simple test program to demonstrate this:

```kotlin
fun main() {
    val client = ModelClientV2.builder().url("http://localhost:28101/v2").build()
    runBlocking {

        val time = measureTime {
            try {
                val res = client.query(BranchReference(RepositoryId("e3_1.2_solution"), "develop-0.52")) { root ->
                    NodeReference("doesnotexist").asMono().resolve()
                }
            } catch (e: Exception) {
                println("Error: $e")
            }
        }

        println("Took: $time")

    }
}
```

Takes 16 seconds for me to error out.

Looking at the new call logging in the model-server, I see the reason:

```[model-server]             | 12:36:37.468 [eventLoopGroupProxy-4-1] INFO  ktor.application - 500 Internal Server Error: POST - /v2/repositories/e3_1.2_solution/branches/develop-0.52/query in 386ms [Remote host: '7dcad08ce605', User agent: 'Ktor client']
[model-server]             | 12:36:40.265 [eventLoopGroupProxy-4-2] DEBUG o.m.modelql.server.ModelQLServer - query: it<1002>.map {<doesnotexist>.resolve()}
[model-server]             | 12:36:40.267 [eventLoopGroupProxy-4-2] INFO  ktor.application - 500 Internal Server Error: POST - /v2/repositories/e3_1.2_solution/branches/develop-0.52/query in 17ms [Remote host: '7dcad08ce605', User agent: 'Ktor client']
[model-server]             | 12:36:44.304 [eventLoopGroupProxy-4-3] DEBUG o.m.modelql.server.ModelQLServer - query: it<1002>.map {<doesnotexist>.resolve()}
[model-server]             | 12:36:44.306 [eventLoopGroupProxy-4-3] INFO  ktor.application - 500 Internal Server Error: POST - /v2/repositories/e3_1.2_solution/branches/develop-0.52/query in 9ms [Remote host: '7dcad08ce605', User agent: 'Ktor client']
[model-server]             | 12:36:52.360 [eventLoopGroupProxy-4-4] DEBUG o.m.modelql.server.ModelQLServer - query: it<1002>.map {<doesnotexist>.resolve()}
[model-server]             | 12:36:52.363 [eventLoopGroupProxy-4-4] INFO  ktor.application - 500 Internal Server Error: POST - /v2/repositories/e3_1.2_solution/branches/develop-0.52/query in 9ms [Remote host: '7dcad08ce605', User agent: 'Ktor client']
```

Due to the model-server returning a 500 and not a reasonable error code, the retry mechanism in `ModelClient` triggers and wastes time retrying the query.

languitar commented:
The exceptions thrown by a query on the result `ModelFacade.toLocalBranch` have changed. Previously, an `assertNotEmpty()` on an empty stream resulted in an `IllegalArgumentException`. Now, we receive a generic RuntimeException.

languitar commented:
Breaking changes to property access are not declared in the commits. I had to change access by UID in queries to access via PropertyReference to preserve functionality.

languitar commented:
I can see the expected flow error also with a model-server backend:

<details>

```
server error(post http://localhost:32821/v2/repositories/testrepo/branches/testbranch/query: 500 internal server error. text: "server version: 9.1.0-27-gfab2fe6-snapshot
com.google.common.util.concurrent.uncheckedexecutionexception: java.lang.runtimeexception: the flow for the query input step is expected to be created by the query
	at com.google.common.cache.localcache$segment.get(localcache.java:2085)
	at com.google.common.cache.localcache.get(localcache.java:4017)
	at com.google.common.cache.localcache$localmanualcache.get(localcache.java:4898)
	at org.modelix.model.server.handlers.inmemorymemoizationpersistence$memoizerimpl.memoize(modelreplicationserver.kt:487)
	at org.modelix.modelql.core.memoizingstep.createflow$lambda$0(memoizingstep.kt:46)
	at com.badoo.reaktive.observable.mapkt$map$1$1.onnext(map.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.switchifemptykt$switchifempty$$inlined$observable$1.subscribe(various.kt:28)
	at com.badoo.reaktive.observable.switchifemptykt$switchifempty$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.zipkt$zip$$inlined$observable$1.subscribe(various.kt:48)
	at com.badoo.reaktive.observable.zipkt$zip$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapobserver.subscribeinner(flatmap.kt:82)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:73)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapobserver.subscribeinner(flatmap.kt:82)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:73)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapobserver.subscribeinner(flatmap.kt:82)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:73)
	at com.badoo.reaktive.observable.observablebyemitterkt$onsubscribeobservable$1.onnext(observablebyemitter.kt:27)
	at com.badoo.reaktive.single.asobservablekt$asobservable$1$1.onsuccess(asobservable.kt:22)
	at com.badoo.reaktive.single.singlebyemitterkt$onsubscribesingle$1.onsuccess(singlebyemitter.kt:27)
	at com.badoo.reaktive.single.mapkt$map$1$1.onsuccess(map.kt:21)
	at com.badoo.reaktive.single.singlebyemitterkt$onsubscribesingle$1.onsuccess(singlebyemitter.kt:27)
	at com.badoo.reaktive.single.mapkt$map$1$1.onsuccess(map.kt:21)
	at com.badoo.reaktive.single.variouskt$singleof$$inlined$singleunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.single.variouskt$singleof$$inlined$singleunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:40)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:39)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:18)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.coroutinesinterop.observableasflowkt$asflow$1.invokesuspend(observableasflow.kt:39)
	at com.badoo.reaktive.coroutinesinterop.observableasflowkt$asflow$1.invoke(observableasflow.kt)
	at com.badoo.reaktive.coroutinesinterop.observableasflowkt$asflow$1.invoke(observableasflow.kt)
	at kotlinx.coroutines.flow.channelflowbuilder.collectto$suspendimpl(builders.kt:316)
	at kotlinx.coroutines.flow.channelflowbuilder.collectto(builders.kt)
	at kotlinx.coroutines.flow.internal.channelflow$collecttofun$1.invokesuspend(channelflow.kt:56)
	at kotlin.coroutines.jvm.internal.basecontinuationimpl.resumewith(continuationimpl.kt:33)
	at kotlinx.coroutines.dispatchedtask.run(dispatchedtask.kt:101)
	at kotlinx.coroutines.eventloopimplbase.processnextevent(eventloop.common.kt:263)
	at kotlinx.coroutines.blockingcoroutine.joinblocking(builders.kt:95)
	at kotlinx.coroutines.builderskt__builderskt.runblocking(builders.kt:69)
	at kotlinx.coroutines.builderskt.runblocking(unknown source)
	at kotlinx.coroutines.builderskt__builderskt.runblocking$default(builders.kt:47)
	at kotlinx.coroutines.builderskt.runblocking$default(unknown source)
	at org.modelix.modelql.server.modelqlserver$companion.handlecall$lambda$4(modelqlserver.kt:93)
	at org.modelix.model.api.treeasbranch.computeread$lambda$0(treeasbranch.kt:39)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.model.api.roleaccesscontext.runwith(irole.kt:91)
	at org.modelix.model.api.treeasbranch.computeread(treeasbranch.kt:39)
	at org.modelix.model.area.parea.executeread$lambda$0(parea.kt:62)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.model.api.inoderesolutionscope$companion.runwithadditionalscope(inoderesolutionscope.kt:52)
	at org.modelix.model.api.inoderesolutionscope.runwithadditionalscope(inoderesolutionscope.kt:34)
	at org.modelix.model.api.inoderesolutionscope$companion.ensureincontext(inoderesolutionscope.kt:93)
	at org.modelix.model.area.parea.executeread(parea.kt:62)
	at org.modelix.modelql.server.modelqlserver$companion.handlecall(modelqlserver.kt:107)
	at org.modelix.modelql.server.modelqlserver$companion$handlecall$4.invokesuspend(modelqlserver.kt)
	at kotlin.coroutines.jvm.internal.basecontinuationimpl.resumewith(continuationimpl.kt:33)
	at kotlinx.coroutines.dispatchedtask.run(dispatchedtask.kt:101)
	at io.netty.util.concurrent.abstracteventexecutor.runtask(abstracteventexecutor.java:173)
	at io.netty.util.concurrent.abstracteventexecutor.safeexecute(abstracteventexecutor.java:166)
	at io.netty.util.concurrent.singlethreadeventexecutor.runalltasks(singlethreadeventexecutor.java:469)
	at io.netty.channel.nio.nioeventloop.run(nioeventloop.java:569)
	at io.netty.util.concurrent.singlethreadeventexecutor$4.run(singlethreadeventexecutor.java:994)
	at io.netty.util.internal.threadexecutormap$2.run(threadexecutormap.java:74)
	at io.ktor.server.netty.eventloopgroupproxy$companion.create$lambda$1$lambda$0(nettyapplicationengine.kt:296)
	at io.netty.util.concurrent.fastthreadlocalrunnable.run(fastthreadlocalrunnable.java:30)
	at java.base/java.lang.thread.run(thread.java:829)
caused by: java.lang.runtimeexception: the flow for the query input step is expected to be created by the query
	at org.modelix.modelql.core.queryinput.createflow(query.kt:496)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.unboundquery.asflow$lambda$17(query.kt:363)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:71)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:40)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:39)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:18)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.base.subscribesafekt.subscribesafe(subscribesafe.kt:7)
	at com.badoo.reaktive.single.subscribekt.subscribe(subscribe.kt:35)
	at com.badoo.reaktive.single.subscribekt.subscribe$default(subscribe.kt:15)
	at org.modelix.streams.streamextensionskt.getsynchronous(streamextensions.kt:95)
	at org.modelix.model.server.handlers.inmemorymemoizationpersistence$memoizerimpl.memoize$lambda$1$lambda$0(modelreplicationserver.kt:488)
	at com.google.common.cache.localcache$localmanualcache$1.load(localcache.java:4903)
	at com.google.common.cache.localcache$loadingvaluereference.loadfuture(localcache.java:3574)
	at com.google.common.cache.localcache$segment.loadsync(localcache.java:2316)
	at com.google.common.cache.localcache$segment.lockedgetorload(localcache.java:2189)
	at com.google.common.cache.localcache$segment.get(localcache.java:2079)
	... 132 more
"
io.ktor.client.plugins.serverresponseexception: server error(post http://localhost:32821/v2/repositories/testrepo/branches/testbranch/query: 500 internal server error. text: "server version: 9.1.0-27-gfab2fe6-snapshot
com.google.common.util.concurrent.uncheckedexecutionexception: java.lang.runtimeexception: the flow for the query input step is expected to be created by the query
	at com.google.common.cache.localcache$segment.get(localcache.java:2085)
	at com.google.common.cache.localcache.get(localcache.java:4017)
	at com.google.common.cache.localcache$localmanualcache.get(localcache.java:4898)
	at org.modelix.model.server.handlers.inmemorymemoizationpersistence$memoizerimpl.memoize(modelreplicationserver.kt:487)
	at org.modelix.modelql.core.memoizingstep.createflow$lambda$0(memoizingstep.kt:46)
	at com.badoo.reaktive.observable.mapkt$map$1$1.onnext(map.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.filterkt$filter$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.switchifemptykt$switchifempty$$inlined$observable$1.subscribe(various.kt:28)
	at com.badoo.reaktive.observable.switchifemptykt$switchifempty$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.zipkt$zip$$inlined$observable$1.subscribe(various.kt:48)
	at com.badoo.reaktive.observable.zipkt$zip$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.mapkt$map$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapobserver.subscribeinner(flatmap.kt:82)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:73)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapobserver.subscribeinner(flatmap.kt:82)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:73)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapobserver.subscribeinner(flatmap.kt:82)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:73)
	at com.badoo.reaktive.observable.observablebyemitterkt$onsubscribeobservable$1.onnext(observablebyemitter.kt:27)
	at com.badoo.reaktive.single.asobservablekt$asobservable$1$1.onsuccess(asobservable.kt:22)
	at com.badoo.reaktive.single.singlebyemitterkt$onsubscribesingle$1.onsuccess(singlebyemitter.kt:27)
	at com.badoo.reaktive.single.mapkt$map$1$1.onsuccess(map.kt:21)
	at com.badoo.reaktive.single.singlebyemitterkt$onsubscribesingle$1.onsuccess(singlebyemitter.kt:27)
	at com.badoo.reaktive.single.mapkt$map$1$1.onsuccess(map.kt:21)
	at com.badoo.reaktive.single.variouskt$singleof$$inlined$singleunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.single.variouskt$singleof$$inlined$singleunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:40)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:39)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:18)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.coroutinesinterop.observableasflowkt$asflow$1.invokesuspend(observableasflow.kt:39)
	at com.badoo.reaktive.coroutinesinterop.observableasflowkt$asflow$1.invoke(observableasflow.kt)
	at com.badoo.reaktive.coroutinesinterop.observableasflowkt$asflow$1.invoke(observableasflow.kt)
	at kotlinx.coroutines.flow.channelflowbuilder.collectto$suspendimpl(builders.kt:316)
	at kotlinx.coroutines.flow.channelflowbuilder.collectto(builders.kt)
	at kotlinx.coroutines.flow.internal.channelflow$collecttofun$1.invokesuspend(channelflow.kt:56)
	at kotlin.coroutines.jvm.internal.basecontinuationimpl.resumewith(continuationimpl.kt:33)
	at kotlinx.coroutines.dispatchedtask.run(dispatchedtask.kt:101)
	at kotlinx.coroutines.eventloopimplbase.processnextevent(eventloop.common.kt:263)
	at kotlinx.coroutines.blockingcoroutine.joinblocking(builders.kt:95)
	at kotlinx.coroutines.builderskt__builderskt.runblocking(builders.kt:69)
	at kotlinx.coroutines.builderskt.runblocking(unknown source)
	at kotlinx.coroutines.builderskt__builderskt.runblocking$default(builders.kt:47)
	at kotlinx.coroutines.builderskt.runblocking$default(unknown source)
	at org.modelix.modelql.server.modelqlserver$companion.handlecall$lambda$4(modelqlserver.kt:93)
	at org.modelix.model.api.treeasbranch.computeread$lambda$0(treeasbranch.kt:39)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.model.api.roleaccesscontext.runwith(irole.kt:91)
	at org.modelix.model.api.treeasbranch.computeread(treeasbranch.kt:39)
	at org.modelix.model.area.parea.executeread$lambda$0(parea.kt:62)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.model.api.inoderesolutionscope$companion.runwithadditionalscope(inoderesolutionscope.kt:52)
	at org.modelix.model.api.inoderesolutionscope.runwithadditionalscope(inoderesolutionscope.kt:34)
	at org.modelix.model.api.inoderesolutionscope$companion.ensureincontext(inoderesolutionscope.kt:93)
	at org.modelix.model.area.parea.executeread(parea.kt:62)
	at org.modelix.modelql.server.modelqlserver$companion.handlecall(modelqlserver.kt:107)
	at org.modelix.modelql.server.modelqlserver$companion$handlecall$4.invokesuspend(modelqlserver.kt)
	at kotlin.coroutines.jvm.internal.basecontinuationimpl.resumewith(continuationimpl.kt:33)
	at kotlinx.coroutines.dispatchedtask.run(dispatchedtask.kt:101)
	at io.netty.util.concurrent.abstracteventexecutor.runtask(abstracteventexecutor.java:173)
	at io.netty.util.concurrent.abstracteventexecutor.safeexecute(abstracteventexecutor.java:166)
	at io.netty.util.concurrent.singlethreadeventexecutor.runalltasks(singlethreadeventexecutor.java:469)
	at io.netty.channel.nio.nioeventloop.run(nioeventloop.java:569)
	at io.netty.util.concurrent.singlethreadeventexecutor$4.run(singlethreadeventexecutor.java:994)
	at io.netty.util.internal.threadexecutormap$2.run(threadexecutormap.java:74)
	at io.ktor.server.netty.eventloopgroupproxy$companion.create$lambda$1$lambda$0(nettyapplicationengine.kt:296)
	at io.netty.util.concurrent.fastthreadlocalrunnable.run(fastthreadlocalrunnable.java:30)
	at java.base/java.lang.thread.run(thread.java:829)
caused by: java.lang.runtimeexception: the flow for the query input step is expected to be created by the query
	at org.modelix.modelql.core.queryinput.createflow(query.kt:496)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.transformingstep.createflow(istep.kt:175)
	at org.modelix.modelql.core.flowinstantiationcontext.getorcreateflow(istep.kt:56)
	at org.modelix.modelql.core.unboundquery.asflow$lambda$17(query.kt:363)
	at com.badoo.reaktive.observable.flatmapobserver.onnext(flatmap.kt:71)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:21)
	at com.badoo.reaktive.observable.variouskt$observableof$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:27)
	at com.badoo.reaktive.observable.flatmapkt$flatmap$$inlined$observable$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.asobservablekt$asobservable$$inlined$observable$1.subscribe(various.kt:13)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:40)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$1$1.invoke(synchronouspipeline.kt:39)
	at org.modelix.kotlin.utils.contextvalue.computewith(contextvalue.jvm.kt:31)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:18)
	at org.modelix.streams.synchronouspipelinekt$endofsynchronouspipeline$$inlined$observableunsafe$1.subscribe(various.kt:13)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.observable.collectkt$collect$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:25)
	at com.badoo.reaktive.single.mapkt$map$$inlined$single$1.subscribe(various.kt:13)
	at com.badoo.reaktive.base.subscribesafekt.subscribesafe(subscribesafe.kt:7)
	at com.badoo.reaktive.single.subscribekt.subscribe(subscribe.kt:35)
	at com.badoo.reaktive.single.subscribekt.subscribe$default(subscribe.kt:15)
	at org.modelix.streams.streamextensionskt.getsynchronous(streamextensions.kt:95)
	at org.modelix.model.server.handlers.inmemorymemoizationpersistence$memoizerimpl.memoize$lambda$1$lambda$0(modelreplicationserver.kt:488)
	at com.google.common.cache.localcache$localmanualcache$1.load(localcache.java:4903)
	at com.google.common.cache.localcache$loadingvaluereference.loadfuture(localcache.java:3574)
	at com.google.common.cache.localcache$segment.loadsync(localcache.java:2316)
	at com.google.common.cache.localcache$segment.lockedgetorload(localcache.java:2189)
	at com.google.common.cache.localcache$segment.get(localcache.java:2079)
	... 132 more
"
	at kotlinx.coroutines.coroutinescopekt.coroutinescope(coroutinescope.kt:259)
	at kotlinx.coroutines.flow.internal.channelflow.collect(channelflow.kt:0)
	at kotlinx.coroutines.flow.flowkt__reducekt.single(reduce.kt:53)
	at kotlinx.coroutines.flow.flowkt.single(unknown source)
	at org.modelix.streams.streamextensionskt.getsuspending(streamextensions.kt:110)
	at org.modelix.modelql.core.monoboundquery.execute(query.kt:82)
	at org.modelix.modelql.untyped.untypedmodelqlkt.query(untypedmodelql.kt:80)
	at org.modelix.modelql.client.modelqlclient.query(modelqlclient.kt:55)
	at org.modelix.model.client2.modelclientv2.query(modelclientv2.kt:430)
	at redacted.adapters.modelql.modelserverqueryable.query(modelserverqueryable.kt:25)
	at redacted.adapters.modelserver.somerepo.createcountcache(somerepo.kt:152)
	at redacted.adapters.modelserver.somerepo$getcountcache$2.invokesuspend(somerepo.kt:92)
	at io.github.reactivecircus.cache4k.realcache$get$3.invokesuspend(realcache.kt:109)
	at io.github.reactivecircus.cache4k.keyedsynchronizer.synchronizedfor(keyedsynchronizer.kt:24)
	at io.github.reactivecircus.cache4k.realcache.get(realcache.kt:98)
	at redacted.adapters.modelserver.somerepo.getcountcache(somerepo.kt:85)
	at redacted.adapters.modelserver.somerepo.countcomponentsoftypes(somerepo.kt:196)
	at redacted.adapters.modelserver.somerepo$invalidates the cache when the version changes$1.invokesuspend(somerepo.kt:133)
	at kotlinx.coroutines.test.testbuilderskt__testbuilderskt$runtest$2$1$1.invokesuspend(testbuilders.kt:318)
	at .(decoroutinator-boundary)
	at io.ktor.client.plugins.defaultresponsevalidationkt$adddefaultresponsevalidation$1$1.invokesuspend(defaultresponsevalidation.kt:55)
	at dev.reformator.stacktracedecoroutinator.common.internal.decoroutinatorspecimpl.resumenext(utils-common.kt:156)
	at io.ktor.client.plugins.httpcallvalidator.validateresponse(httpcallvalidator.kt:51)
	at io.ktor.client.plugins.httpcallvalidator$companion$install$3.invokesuspend(httpcallvalidator.kt:152)
	at io.ktor.client.plugins.httpsend$interceptedsender.execute(httpsend.kt:116)
	at io.ktor.client.plugins.httpsend$plugin$install$1.invokesuspend(httpsend.kt:104)
	at dev.reformator.stacktracedecoroutinator.common.internal.awakenerkt.callspecmethods(awakener.kt:80)
	at dev.reformator.stacktracedecoroutinator.common.internal.awakenerkt.awake(awakener.kt:32)
	at dev.reformator.stacktracedecoroutinator.common.internal.provider.awakebasecontinuation(provider-impl.kt:43)
	at dev.reformator.stacktracedecoroutinator.provider.decoroutinatorproviderapikt.awakebasecontinuation(provider-api.kt:47)
	at kotlin.coroutines.jvm.internal.basecontinuationimpl.resumewith(continuationimpl.kt)
	at kotlinx.coroutines.internal.dispatchedcontinuation.resumewith(dispatchedcontinuation.kt:197)
	at io.ktor.utils.io.internal.cancellablereusablecontinuation.resumewith(cancellablereusablecontinuation.kt:93)
	at io.ktor.utils.io.bytebufferchannel.resumereadop(bytebufferchannel.kt:2104)
	at io.ktor.utils.io.bytebufferchannel.tryterminate$ktor_io(bytebufferchannel.kt:390)
	at io.ktor.utils.io.bytebufferchannel.close(bytebufferchannel.kt:135)
	at io.ktor.utils.io.coroutineskt$launchchannel$1.invoke(coroutines.kt:158)
	at io.ktor.utils.io.coroutineskt$launchchannel$1.invoke(coroutines.kt:157)
	at kotlinx.coroutines.invokeoncompletion.invoke(jobsupport.kt:1534)
	at kotlinx.coroutines.jobsupport.notifycompletion(jobsupport.kt:1625)
	at kotlinx.coroutines.jobsupport.completestatefinalization(jobsupport.kt:316)
	at kotlinx.coroutines.jobsupport.finalizefinishingstate(jobsupport.kt:233)
	at kotlinx.coroutines.jobsupport.trymakecompletingslowpath(jobsupport.kt:946)
	at kotlinx.coroutines.jobsupport.trymakecompleting(jobsupport.kt:894)
	at kotlinx.coroutines.jobsupport.makecompletingonce$kotlinx_coroutines_core(jobsupport.kt:859)
	at kotlinx.coroutines.abstractcoroutine.resumewith(abstractcoroutine.kt:98)
	at dev.reformator.stacktracedecoroutinator.common.internal.awakenerkt.awake(awakener.kt:47)
	at dev.reformator.stacktracedecoroutinator.common.internal.provider.awakebasecontinuation(provider-impl.kt:43)
	at dev.reformator.stacktracedecoroutinator.provider.decoroutinatorproviderapikt.awakebasecontinuation(provider-api.kt:47)
	at kotlin.coroutines.jvm.internal.basecontinuationimpl.resumewith(continuationimpl.kt)
	at kotlinx.coroutines.dispatchedtask.run(dispatchedtask.kt:101)
	at kotlinx.coroutines.internal.limiteddispatcher$worker.run(limiteddispatcher.kt:113)
	at kotlinx.coroutines.scheduling.taskimpl.run(tasks.kt:89)
	at kotlinx.coroutines.scheduling.coroutinescheduler.runsafely(coroutinescheduler.kt:589)
	at kotlinx.coroutines.scheduling.coroutinescheduler$worker.executetask(coroutinescheduler.kt:823)
	at kotlinx.coroutines.scheduling.coroutinescheduler$worker.runworker(coroutinescheduler.kt:720)
	at kotlinx.coroutines.scheduling.coroutinescheduler$worker.run(coroutinescheduler.kt:707)
```

</details>

languitar commented:
The exceptions with "the flow for the query input step is expected to be created by the query" were caused by access a variable from inside `memoize` that was defined outside. Can we get a clearer exception message for this case?

And one last thing: a `findAll` variant accepting a Flux of keys to find would be nice.

slisson commented:
> In the current state, at least one commit needs to be marked as a breaking commit. When using `ModelFacade.toLocalBranch(ModelFacade.newLocalTree())` and then resolving references via ModelQL on the resulting branch, resolution fails due to the lack of an INodeResolutionScope being set in code that worked before.

Fixed that. Also deprecated `ModelFacade.toLocalBranch` and added documentation explaining when and when not to use `TreePointer`.

slisson commented:
> One more thing: I've built a docker image of the model-server from this branch. With postgres persistence, the image does not start:

Can't reproduce that.

slisson commented:
> Trying to resolve non-existing nodes via ModelQl takes ages. It was never fast before, but it feels even longer now. Here's a simple test program to demonstrate this:
> Due to the model-server returning a 500 and not a reasonable error code, the retry mechanism in `ModelClient` triggers and wastes time retrying the query.

Now returns a 404.

languitar commented on modelql-client/src/jvmTest/kotlin/org/modelix/modelql/client/ModelQLClientTest.kt:
```diff
@@ -279,4 +283,15 @@ class ModelQLClientTest {
     }
+
+    @Test
+    fun `resolving a non-existing node reference returns 404`() = runTest { httpClient ->
```
We should probably document that explicitly in the OpenAPI specification.

slisson replied:
Updated the OpenAPI spec.

languitar reviewed (approved):
One last nit. Apart from that, let's give that a try.

slisson commented:
> Breaking changes to property access are not declared in the commits. I had to change access by UID in queries to access via PropertyReference to preserve functionality.

The reason was that the unit tests used named based role persistence. I changed all default property values for `useRoleIds` to `true` and marked that as a breaking change. Named based persistence shouldn't be used for new models, but if anyone still relies on it, the code can simply be migrated by providing an explicit `false` value for it.

slisson commented:
> And one last thing: a `findAll` variant accepting a Flux of keys to find would be nice.

Added.

slisson commented:
> The exceptions with "the flow for the query input step is expected to be created by the query" were caused by accessing a variable from inside `memoize` that was defined outside. Can we get a clearer exception message for this case?

Changed the exception message to `"Unsupported cross-query usage of $this in ${owner.query}"`

slisson commented:
:tada: This PR is included in version 10.0.0 :tada:

The release is available on [GitHub release](https://github.com/modelix/modelix.core/releases/tag/10.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
